### PR TITLE
[relay]: Prevent Buffer Overrun Of Malformed DHCP Packet

### LINKED ISSUE #6052: DHCP relay forwards incorrect DHCP client packet to DHCP server(in case of DHCP packet without payload)
[dump_on_dhcp_server_side.pcap.gz](https://github.com/Azure/sonic-buildimage/files/5603026/dump_on_dhcp_server_side.pcap.gz)
[dump_on_dut_dhcp_client_vlan.pcap.gz](https://github.com/Azure/sonic-buildimage/files/5603029/dump_on_dut_dhcp_client_vlan.pcap.gz)

**DHCP relay forwards incorrect DHCP client packet to DHCP server(in case of DHCP packet without payload)**

DHCP relay forwards incorrect DHCP client packet to DHCP server(in case of DHCP packet without payload)
Issue happens in case when send DHCP packet via scapy without DHCP payload.
Packet example:  Ether(src="0c:42:a1:17:e6:fd",dst="ff:ff:ff:ff:ff:ff")/IP(src="0.0.0.0",dst="255.255.255.255")/UDP(sport=68,dport=67)/BOOTP(chaddr='0c42a117e6fd'.decode('hex'),xid=RandInt())/DHCP()

**Setup**:
SONiC device has 2 connected Linux hosts(host A, host B)
Host A connected from port enp4s0f1 to SONiC port Ethernet124
Host B connected from port enp6s0f1 to SONiC port Ethernet60
Host A acts as DCHP client, 
Host B acts as DCHP server and has next network configuration:
sudo ip addr add 69.0.0.2/24 dev enp6s0f1
sudo ip route add 69.0.1.0/24 via 69.0.0.1

**Steps to reproduce the issue:**
1. Configure DUT:
sudo config interface ip add Ethernet60 69.0.0.1/24
sudo config vlan add 690
sudo config vlan member add --untagged 690 Ethernet124
sudo config interface ip add Vlan690 69.0.1.1/24
sudo config vlan dhcp_relay add 690 69.0.0.2
2. Send from DHCP client empty DHCP packet using scapy:
pkt = Ether(src="0c:42:a1:17:e6:fd",dst="ff:ff:ff:ff:ff:ff")/IP(src="0.0.0.0",dst="255.255.255.255")/UDP(sport=68,dport=67)/BOOTP(chaddr='0c42a117e6fd'.decode('hex'),xid=RandInt())/DHCP()
sendp(pkt, iface='enp4s0f1')
3. Check packet to DHCP server side

**Describe the results you received:**
DHCP packet arrived to DHCP server, but it modified and have some unexpected data. For example in packet added option 12 - with some hostname, also added option 50 - with requested IP - which is not in DHCP server subnet and other.
See attached PCAP files(use: "tar -xf file_name.pcap..gz"  for unpack).

**Describe the results you expected:**
DHCP packet forwarded to DHCP server without modification, only DHCP relay option 82 should be added.

**Additional information you deem important (e.g. issue happens only occasionally):**
Issue reproduce only in case of DHCP packet without payload.

 ```
SONiC Software Version: SONiC.201911.30-dirty-20201124.120644
Distribution: Debian 9.13
Kernel: 4.9.0-11-2-amd64
Build commit: ac5117f2
Build date: Tue Nov 24 11:57:24 UTC 2020
Built by: sw-r2d2-bot@r-build-sonic-ci02

Platform: x86_64-mlnx_msn3700-r0
HwSKU: ACS-MSN3700
ASIC: mellanox
Serial Number: MT1949X06182
Uptime: 12:27:39 up 13:35,  1 user,  load average: 1.88, 1.52, 1.14

Docker images:
REPOSITORY                    TAG                               IMAGE ID            SIZE
docker-syncd-mlnx             201911.30-dirty-20201124.120644   e758fd3a3240        399MB
docker-syncd-mlnx             latest                            e758fd3a3240        399MB
docker-sonic-telemetry        201911.30-dirty-20201124.120644   4354e24f368d        353MB
docker-sonic-telemetry        latest                            4354e24f368d        353MB
docker-router-advertiser      201911.30-dirty-20201124.120644   74d4310804b7        290MB
docker-router-advertiser      latest                            74d4310804b7        290MB
docker-platform-monitor       201911.30-dirty-20201124.120644   2904669937a5        665MB
docker-platform-monitor       latest                            2904669937a5        665MB
docker-fpm-frr                201911.30-dirty-20201124.120644   804893596a66        335MB
docker-fpm-frr                latest                            804893596a66        335MB
docker-teamd                  201911.30-dirty-20201124.120644   43862681ed22        315MB
docker-teamd                  latest                            43862681ed22        315MB
docker-lldp-sv2               201911.30-dirty-20201124.120644   96bf1ad6f224        312MB
docker-lldp-sv2               latest                            96bf1ad6f224        312MB
docker-dhcp-relay             201911.30-dirty-20201124.120644   90aa45d7faf6        300MB
docker-dhcp-relay             latest                            90aa45d7faf6        300MB
docker-database               201911.30-dirty-20201124.120644   eaac4565aec3        289MB
docker-database               latest                            eaac4565aec3        289MB
docker-snmp-sv2               201911.30-dirty-20201124.120644   4046aec714f3        348MB
docker-snmp-sv2               latest                            4046aec714f3        348MB
docker-orchagent              201911.30-dirty-20201124.120644   7bb60251023b        333MB
docker-orchagent              latest                            7bb60251023b        333MB
docker-sflow                  201911.30-dirty-20201124.120644   40dcd66b4af4        315MB
docker-sflow                  latest                            40dcd66b4af4        315MB
docker-nat                    201911.30-dirty-20201124.120644   40757ed6bbfb        316MB
docker-nat                    latest                            40757ed6bbfb        316MB
docker-sonic-mgmt-framework   201911.30-dirty-20201124.120644   354cc3e4d510        427MB
docker-sonic-mgmt-framework   latest                            354cc3e4d510        427MB
 ```

### PR DESCRIPTION
**Pull request by tahmed-dev**
The add/strip relay agent options do not take into account the buffer
length and so it is possible to overrun the buffer. The issue will
result in contents from previous packet being added to the current one.

closes #6052  
resolves #6052 

signed-off-by: Tamer Ahmed <tamer.ahmed@microsoft.com>

**- Why I did it**

**- How I did it**

**- How to verify it**

**- Which release branch to backport (provide reason below if selected)**

- [ ] 201811
- [ ] 201911
- [ ] 202006

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

File: src/isc-dhcp/patch/0009-relay-Prevent-Buffer-Overrun.patch [hunk 1]
```diff
@@ -0,0 +1,30 @@
+From 2fcfd6277ab5105f970a4e2872a0040bc8fc8744 Mon Sep 17 00:00:00 2001
+From: Tamer Ahmed <tamer.ahmed@microsoft.com>
+Date: Sat, 28 Nov 2020 16:28:37 -0800
+Subject: [PATCH] [relay] Prevent Buffer Overrun
+
+The add/strip relay agent options do not take into account the buffer
+length and so it is possible to overrun the buffer. The issue will
+result in contents from previous packet being aded to the current one.
```
Comment (jleveque) [inline src/isc-dhcp/patch/0009-relay-Prevent-Buffer-Overrun.patch:8]: Typo: s/aded/added/

**Review by jleveque — CHANGES_REQUESTED** [review]

**Comment by tahmed-dev** [comment]
retest mellanox please

**Comment by tahmed-dev** [comment]
retest mellanox please

**Comment by tahmed-dev** [comment]
retest this please

**Review by jleveque — APPROVED** [review]
LGTM. Should this patch be upstreamed?

**Comment by tahmed-dev** [comment]
> LGTM. Should this patch be upstreamed?

I guess so. There could be cross packets information leak.

**Review by trzhang-msft — APPROVED** [review]

**Pull request merged**